pkg/controller/operconfig/ingressconfig_controller.go

```diff
@@ -0,0 +1,96 @@
+package operconfig
+
+import (
+	"context"
+	"log"
+
+	operv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-network-operator/pkg/names"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// newIngressConfigReconciler returns a new reconcile.Reconciler
+func newIngressConfigReconciler(client client.Client) *ReconcileIngressConfigs {
+	return &ReconcileIngressConfigs{client: client}
+}
+
+var _ reconcile.Reconciler = &ReconcileIngressConfigs{}
+
+// ReconcileIngressConfigs watches for updates to ingress controller configuration
+// and sets the network policy related labels on the openshift-host-network namespace
+type ReconcileIngressConfigs struct {
+	client client.Client
+}
+
+// Reconcile sets the openshift-host-network namespaces' labels as per the
+// endpointPublishingStrategy of the `default` ingress controller object.
+func (r *ReconcileIngressConfigs) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	if request.Namespace == names.IngressControllerNamespace && request.Name == names.DefaultIngressControllerName {
+		log.Printf("Reconciling update to %s/%s\n", request.Namespace, request.Name)
+		ingressControllerConfig := &operv1.IngressController{TypeMeta: metav1.TypeMeta{APIVersion: operv1.GroupVersion.String(), Kind: "IngressController"}}
+		err := r.client.Get(context.TODO(), request.NamespacedName, ingressControllerConfig)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Printf("Ingress Controller configuration %s was deleted", request.NamespacedName.String())
+				// Request object not found, could have been deleted after reconcile request.
+				// Owned objects are automatically garbage collected, since we set
+				// the ownerReference (see https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/).
+				// Return and don't requeue
+				return reconcile.Result{}, nil
+			}
+			// Error reading the object - requeue the request.
+			log.Printf("Unable to retrieve IngressController.operator.openshift.io object: %v", err)
+			return reconcile.Result{}, err
+		}
+		labels := map[string]*string{names.PolicyGroupLabel: nil}
+		log.Printf("Ingress Controller Config: %+v", ingressControllerConfig)
+		if ingressControllerConfig.Spec.EndpointPublishingStrategy != nil &&
+			ingressControllerConfig.Spec.EndpointPublishingStrategy.Type == operv1.HostNetworkStrategyType {
+			labelVal := names.PolicyGroupLabelIngressValue
+			labels[names.PolicyGroupLabel] = &labelVal
+		}
+		err = r.setLabelsOnNamespace(names.HostNetworkNamespace, labels)
```

Comment:
Rather than hard-coding the namespace name, it would be better to _retrieve_ the host-network-equivalent namespace with a label selector. In other words, look for `network.openshift.io/policy-group = host-network`

That leaves the door open for third-party network providers to implement this functionality.

Reply:
The `setLabelsOnNamespace` API is only used to set the `policy-group:ingress` label on the namespace that OpenShift's native network plugins will use for host-network traffic. For other providers, they may choose a completely different mechanism to implement network policy for host network traffic. That also potentially means we will expose a vulnerability that anyone who can create a namespace can slap the host-network label on that namespace and have host-network traffic allowed to pods?

Reply:
Labelling namespaces is a privileged operation, so that's not a risk.

The whole point is to allow third-party vendors to hook in to this system; We have a standardized label API; if they conform to that they would be automatically supported without any extra customization.

Reply:
How do we handle multiple namespaces with the `host-network` label on them? Do we label all such namespaces with ingress label? I get a feeling that in the attempt to prematurely optimize, we may open up new cans of worms. If there is a legit case where a third-party vendor wants to implement network policy for host-network pods using the mechanism we provide, then we can make the change being suggested at that point.

Reply:
We would certainly label multiple namespaces - that would be the exact behavior we would expect.

Basically, we're saying that "when the Ingress controller is in host-network mode, we must ensure that the host-network stub namespaces have `policy-group/ingress`. How do we know which namespaces are host-network? We look for the label-selector.

By not hard-coding the namespace name, it also means that we no longer need any code changes to openshift-sdn. Just label `default` with `policy-group/host-network`, and everything else takes care of itself.

Reply:
PTAL at https://github.com/openshift/sdn/pull/262/files. The changes for openshift-sdn aren't too bad. I am still not convinced we need to look up the namespace especially when we don't have a third-party plugin that would use the mechanism we are providing here to enable network policies on host-network traffic. Not to mention that the third-party plugins still need to have code in place in order to use the labels that we are providing on whatever namespace they choose to classify host-network traffic into. So this isn't quite seamless. @danwinship wdyt?

Reply:
> By not hard-coding the namespace name, it also means that we no longer need any code changes to openshift-sdn. Just label `default` with `policy-group/host-network`, and everything else takes care of itself.

The fact that we don't use `default` in openshift-sdn is intentional. (`default` is not supposed to be magical, and while we can't fix it at this point, I don't want to encourage people to think of it that way, especially since it leads to people thinking ovn-kubernetes is broken for not having the same behavior.)


I don't like having CNO modify Namespaces that it didn't create. If a third-party network plugin wants to create their own "host-network" namespace, they can have their own ingress config watcher too.